Xamarin.Essentials/Types/FileProvider.android.cs

```diff
@@ -0,0 +1,102 @@
+﻿using System;
+using Android.App;
+using Android.Content;
+using Android.OS;
+using AndroidEnvironment = Android.OS.Environment;
+
+namespace Xamarin.Essentials
+{
+    [ContentProvider(
+        new[] { "${applicationId}.fileProvider" },
+        Name = "xamarin.essentials.fileProvider",
+        Exported = false,
+        GrantUriPermissions = true)]
+    [MetaData(
+        "android.support.FILE_PROVIDER_PATHS",
+        Resource = "@xml/xamarin_essentials_fileprovider_file_paths")]
+    public class FileProvider : global::Android.Support.V4.Content.FileProvider
+    {
+        internal static bool AlwaysFailExternalMediaAccess { get; set; } = false;
+
+        public static FileProviderLocation TemporaryLocation { get; set; }
+
+        internal static Java.IO.File GetTemporaryDirectory()
+        {
+            var root = GetTemporaryRootDirectory();
+            var dir = new Java.IO.File(root, "2203693cc04e0be7f4f024d5f9499e13");
+            dir.Mkdirs();
+            dir.DeleteOnExit();
+            return dir;
+        }
+
+        internal static Java.IO.File GetTemporaryRootDirectory()
+        {
+            // we want the internal storage
+            if (TemporaryLocation == FileProviderLocation.Internal)
+                return Platform.AppContext.CacheDir;
+
+            // if we need permission, then check
+            // if we can fall back, then don't throw
+            var externalOnly = TemporaryLocation == FileProviderLocation.External;
+            var hasPermission = Platform.HasApiLevel(BuildVersionCodes.Kitkat) || Permissions.EnsureDeclared(PermissionType.WriteExternalStorage, externalOnly);
```

Comment:
Just want to double check the intent of this code.  We are checking if the API level is KitKat, we don't need the declare the permission in the manifest?  I would expect this to look something like if APILevel <= KitKat, but presumably since we don't support below KitKat this is ok.  Am I correct in interpreting the intent here?

Reply:
The opposite if API > kitkat, then no permission. else permission:
> Starting in API level 19, this permission is not required to read/write files in your application-specific directories returned by Context.getExternalFilesDir(String) and Context.getExternalCacheDir().
https://developer.android.com/reference/android/Manifest.permission.html#WRITE_EXTERNAL_STORAGE